File: apps/android/app/src/main/java/com/exponential/app/data/api/CommentsApi.kt
package com.exponential.app.data.api

import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class CreateCommentInput(
    @SerialName("issueId") val issueId: String,
    val body: String,
    // EXP-554: the attachments to link to the new comment. The shared Json has
    // `explicitNulls = false`, so null is OMITTED from the wire body rather
    // than sent as `null` — older servers keep parsing the input.
    @SerialName("attachmentIds") val attachmentIds: List<String>? = null,
)

@Serializable
data class UpdateCommentInput(
    val id: String,
    val body: String,
    // Omitted = attachments untouched (what the MCP tools rely on). An array is
    // the FULL desired set: rows missing from it are hard-deleted server-side.
    @SerialName("attachmentIds") val attachmentIds: List<String>? = null,
)

@Serializable
data class DeleteCommentInput(val id: String)

@Singleton
class CommentsApi @Inject constructor(private val trpc: TrpcClient) {

    suspend fun create(
        accountId: String,
        issueId: String,
        text: String,
        attachmentIds: List<String>? = null,
    ) {
        trpc.mutationUnit(
            accountId,
            path = "comments.create",
            input = CreateCommentInput(issueId, text, attachmentIds),
            inputSerializer = CreateCommentInput.serializer(),
        )
    }

    suspend fun update(
        accountId: String,
        id: String,
        text: String,
        attachmentIds: List<String>? = null,
    ) {
        trpc.mutationUnit(
            accountId,
            path = "comments.update",
            input = UpdateCommentInput(id, text, attachmentIds),
            inputSerializer = UpdateCommentInput.serializer(),
        )
    }

    suspend fun delete(accountId: String, id: String) {
        trpc.mutationUnit(
            accountId,
            path = "comments.delete",
            input = DeleteCommentInput(id),
            inputSerializer = DeleteCommentInput.serializer(),
        )
    }
}

// Extract the `{ "text": "..." }` field from a JSONB comment body stored as
// stringified JSON by Electric. Mirrors getCommentBodyText in the web app.
fun getCommentBodyText(body: String?): String {
    if (body.isNullOrBlank()) return ""
    return try {
        val element = kotlinx.serialization.json.Json.parseToJsonElement(body)
        val obj = element as? kotlinx.serialization.json.JsonObject
        val text = obj?.get("text") as? kotlinx.serialization.json.JsonPrimitive
        text?.content ?: body
    } catch (_: Throwable) {
        body
    }
}

File: apps/ios/ExpCore/Sources/API/CommentsApi.swift
import Foundation

// Comment bodies are plain GFM markdown strings (the server stores them in a
// `text` column; the legacy jsonb `{text}` envelope was dropped in Phase F).
public struct CreateCommentInput: Encodable, Sendable {
    public let issueId: String
    public let body: String
    /// EXP-554 — the attachment rows this comment links (uploaded first through
    /// the REST image/file routes). Optional on the wire: the server defaults it
    /// to `[]`, and Swift's synthesized `encode(to:)` uses `encodeIfPresent`, so
    /// a nil simply never appears in the JSON body.
    public let attachmentIds: [String]?

    public init(issueId: String, body: String, attachmentIds: [String]? = nil) {
        self.issueId = issueId
        self.body = body
        self.attachmentIds = attachmentIds
    }
}

public struct UpdateCommentInput: Encodable, Sendable {
    public let id: String
    public let body: String
    /// The FULL desired set — rows linked to this comment but missing from it
    /// are hard-deleted server-side. Omitted (nil) leaves attachments untouched,
    /// which is what the MCP tools send.
    public let attachmentIds: [String]?

    public init(id: String, body: String, attachmentIds: [String]? = nil) {
        self.id = id
        self.body = body
        self.attachmentIds = attachmentIds
    }
}

public struct DeleteCommentInput: Encodable, Sendable {
    public let id: String

    public init(id: String) {
        self.id = id
    }
}

// The web tRPC handlers return { txId, comment } on create/update and { txId }
// on delete. We don't read txId on iOS — Electric eventually delivers the
// canonical row — so we accept any decodable response shape.
private struct EmptyResult: Decodable {}

public final class CommentsApi: Sendable {
    private let trpc: TrpcClient

    public init(trpc: TrpcClient) {
        self.trpc = trpc
    }

    public func create(
        accountId: String,
        issueId: String,
        text: String,
        attachmentIds: [String]? = nil
    ) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "comments.create",
            input: CreateCommentInput(issueId: issueId, body: text, attachmentIds: attachmentIds)
        )
    }

    public func update(
        accountId: String,
        id: String,
        text: String,
        attachmentIds: [String]? = nil
    ) async throws {
        let _: EmptyResult = try await trpc.mutation(
            accountId: accountId,
            path: "comments.update",
            input: UpdateCommentInput(id: id, body: text, attachmentIds: attachmentIds)
        )
    }

    public func delete(accountId: String, id: String) async throws {
        try await trpc.mutationVoid(accountId: accountId, path: "comments.delete", input: DeleteCommentInput(id: id))
    }
}
